source/date_time.c

```diff
@@ -0,0 +1,826 @@
+                } else if (isdigit(c)) {
+                    parsed_time->tm_year = parsed_time->tm_year * 10 + (c - '0');
+                } else {
+                    error = true;
+                }
+                break;
+            case 5:
+                if (c == ':' && index - state_start_index == 2) {
+                    state = 6;
+                    state_start_index = index + 1;
+                } else if (isdigit(c)) {
+                    parsed_time->tm_hour = parsed_time->tm_hour * 10 + (c - '0');
+                } else {
+                    error = true;
+                }
+                break;
+            case 6:
+                if (c == ':' && index - state_start_index == 2) {
+                    state = 7;
+                    state_start_index = index + 1;
+                } else if (isdigit(c)) {
+                    parsed_time->tm_min = parsed_time->tm_min * 10 + (c - '0');
+                } else {
+                    error = true;
+                }
+                break;
+            case 7:
+                if (isspace(c) && index - state_start_index == 2) {
+                    state = 8;
+                    state_start_index = index + 1;
+                } else if (isdigit(c)) {
+                    parsed_time->tm_sec = parsed_time->tm_sec * 10 + (c - '0');
+                } else {
+                    error = true;
+                }
+                break;
+            case 8:
+                if (isalpha(c) && (index - state_start_index) < 5) {
+                    dt->tz[index - state_start_index] = c;
+                }
+
+                break;
+        }
+
+        index++;
+    }
+
+    if (dt->tz[0] != 0) {
+        dt->utc_assumed = is_utc_time_zone(dt->tz);
+    }
+
+    return error || state != finalState ? AWS_OP_ERR : AWS_OP_SUCCESS;
+}
+
+int aws_date_time_init_from_str(
+    struct aws_date_time *dt,
+    const struct aws_byte_buf *date_str,
+    enum aws_date_format fmt) {
+    if (date_str->len > 100) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    AWS_ZERO_STRUCT(*dt);
+
+    struct tm parsed_time;
+    AWS_ZERO_STRUCT(parsed_time);
+    bool successfully_parsed = false;
+    if (fmt == AWS_DATE_FORMAT_ISO_8601 || fmt == AWS_DATE_FORMAT_AUTO_DETECT) {
+        if (!s_parse_iso_8601(date_str, &parsed_time)) {
+            dt->utc_assumed = true;
+            successfully_parsed = true;
+        }
+    }
+
+    if (fmt == AWS_DATE_FORMAT_RFC822 || (fmt == AWS_DATE_FORMAT_AUTO_DETECT && !successfully_parsed)) {
+        if (!s_parse_rfc_822(date_str, &parsed_time, dt)) {
+            successfully_parsed = true;
+        }
+    }
+
+    if (!successfully_parsed) {
+        return aws_raise_error(AWS_ERROR_INVALID_DATE_STR);
+    }
+
+    if (dt->utc_assumed) {
+        dt->timestamp = aws_mk_gm_time(&parsed_time);
+    } else {
+        dt->timestamp = mktime(&parsed_time);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_date_time_to_local_time_str(
```

Comment:
Don't we usually put output/destination buffers as the first argument?

Reply:
last argument, which it is here.